src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,495 @@
+  χ⁻¹ a = χ (ring.inverse a) :=
+begin
+  by_cases ha : is_unit a,
+  { rw [inv_apply_eq_inv],
+    have h := is_unit.map χ ha,
+    apply_fun ((*) (χ a)) using is_unit.mul_right_injective h,
+    rw [ring.mul_inverse_cancel _ h, ← map_mul, ring.mul_inverse_cancel _ ha, mul_char.map_one], },
+  { revert ha, nontriviality R, intro ha, -- `nontriviality R` by itself doesn't do it
+    rw [map_nonunit _ ha, ring.inverse_non_unit a ha, mul_char.map_zero χ], },
+end
+
+/-- When the domain is a field, we can use the field inverse instead. -/
+lemma inv_apply' {R : Type u} [field R] (χ : mul_char R R') (a : R) : χ⁻¹ a = χ a⁻¹ :=
+(inv_apply χ a).trans $ congr_arg _ (ring.inverse_eq_inv a)
+
+/-- The product of a character with its inverse is the trivial character. -/
+@[simp]
+lemma inv_mul (χ : mul_char R R') : χ⁻¹ * χ = 1 :=
+begin
+  ext x,
+  rw [coe_to_fun_mul, pi.mul_apply, inv_apply_eq_inv,
+      ring.inverse_mul_cancel _ (is_unit.map _ x.is_unit), one_apply_coe],
+end
+
+/-- Finally, the commutative group structure on `mul_char R R'`. -/
+noncomputable
+instance comm_group : comm_group (mul_char R R') :=
+{ one := 1,
+  mul := (*),
+  inv := has_inv.inv,
+  mul_left_inv := inv_mul,
+  mul_assoc := by { intros χ₁ χ₂ χ₃, ext a, simp [mul_assoc], },
+  mul_comm := by { intros χ₁ χ₂, ext a, simp [mul_comm], },
+  one_mul := one_mul,
+  mul_one := mul_one, }
+
+/-- If `a` is a unit and `n : ℕ`, then `(χ ^ n) a = (χ a) ^ n`. -/
+lemma pow_apply_coe (χ : mul_char R R') (n : ℕ) (a : Rˣ) :
+  (χ ^ n) a = (χ a) ^ n :=
+begin
+  induction n with n ih,
+  { rw [pow_zero, pow_zero, one_apply_coe], },
+  { rw [pow_succ, pow_succ, mul_apply, ih], },
+end
+
+/-- If `n` is positive, then `(χ ^ n) a = (χ a) ^ n`. -/
+lemma pow_apply' (χ : mul_char R R') {n : ℕ} (hn : 0 < n) (a : R) :
+  (χ ^ n) a = (χ a) ^ n :=
+begin
+  by_cases ha : is_unit a,
+  { exact pow_apply_coe χ n ha.unit, },
+  { rw [map_nonunit (χ ^ n) ha, map_nonunit χ ha, zero_pow hn], },
+end
+
+end mul_char
+
+end group
+
+end definition_and_group
+
+/-!
+### Properties of multiplicative characters
+
+We introduce the properties of being nontrivial or quadratic and prove
+some basic facts about them.
+
+We now assume that domain and target are commutative rings.
+-/
+
+section properties
+
+namespace mul_char
+
+universes u v w
+
+variables {R : Type u} [comm_ring R] {R' : Type v} [comm_ring R'] {R'' : Type w} [comm_ring R'']
+
+/-- A multiplicative character is *nontrivial* if it takes a value `≠ 1` on a unit. -/
+def is_nontrivial (χ : mul_char R R') : Prop := ∃ a : Rˣ, χ a ≠ 1
+
+/-- A multiplicative character is nontrivial iff it is not the trivial character. -/
+lemma is_nontrivial_iff (χ : mul_char R R') : χ.is_nontrivial ↔ χ ≠ 1 :=
+begin
+  split,
+  { intros h₁ h₂,
+    obtain ⟨a, ha⟩ := h₁,
+    rw [h₂, one_apply_coe] at ha,
+    exact ha rfl, },
+  { contrapose!,
+    intro h,
+    change ¬ ∃ a : Rˣ, χ a ≠ 1 at h,
+    push_neg at h,
+    ext a,
+    rw [one_apply_coe a, h a], },
+end
```

Comment:
With `ext_iff`, we can summarize this as:
```suggestion
by simp only [is_nontrivial, ne.def, ext_iff, not_forall, one_apply_coe]
```

Reply:
That's indeed nice and short.